mattermost/mattermost-plugin-apps
# Kubeless support

## Support for serverless Apps running on Kubeless

### Status
- All PR feedback is either addressed, or answered and awaiting confirmation, cc @hanzei @mickmister 
- e2e cypress test is failing because `/examples/go/hello` has been removed, @mickmister will provide a separate PR that removes it cleanly.
- ~~CodeQL job is failing, and I have no explanation for it. @hanzei can you help, please?~~ not failing now

#### Testing:
- Install and configure `kubectl`, `kubeless` CLI
- Create the Kubernetes namespace for Mattermost Apps:
  ```sh
  kubectl create namespace mattermost-kubeless-apps
  ```
- Build `hello-kubeless` bundle
  ```sh
  cd examples/js/hello-kubeless && make dist
  ```
- Provision the App in Kubeless
  ```sh
  appsctl kubeless provision -v ./examples/js/hello-kubeless/dist/bundle.zip
  ```
- Install the App in Mattermost (note, at the moment there is no good place to host the manifest, so installing from a github URL): 
  ```
  /apps install kubeless https://raw.githubusercontent.com/mattermost/mattermost-plugin-apps/kubeless/examples/js/hello-kubeless/manifest.json
  /hello-kubeless send
  ```
- Observe: 
![image](https://user-images.githubusercontent.com/1187448/129294078-088d13f8-47f0-4584-848d-6dffdabd266d.png)

#### Changes
##### Kubeless support
- added `upstream/upkubeless` which implements upstream and provisioning for Kubeless
- added `upstream/serverless.go` to transform between http and serverless requests
- added `/apps install kubeless` command
- **BREAKING** `appsctl provision app` is now `appsctl [aws|kubeless] provision`
- added `appsctl kubeless provision|test` commands
- added `examples/js/hello-kubeless`. Note: `go` apps are not presently supported on Kubeless since it only provides `go1.14` runtime out of the box, and `...-plugin-apps` requires 1.16

##### Other
- Moved `PluginID` from `App` to `Manifest`
- `Upstream`s are now cached and are App-independent
- Slightly refactored service initialization and configuration for `store`, `proxy` mostly to support the changes in `upstream.Upstream`
